README: Improved build and usage instructions. (#75)
godotengine/godot-ios-plugins: README.md

# Godot iOS plugins

> **Note:** iOS plugins are not editor plugins.

Reviewer: Use GitHub admonition syntax:

```suggestion
> [!NOTE]
>
> iOS plugins are not editor plugins.
```

### For Godot 4.x:

> 1. After cloning this repository, go to `godot` submodule and checkout the tag of the engine version you are working with. For example `4.3-stable`.

Reviewer: `git switch` is preferred over `git checkout` nowadays, as it avoids the ambiguity of `git checkout` also being able to work as `git restore`:

```suggestion
1. After cloning this repository, go to `godot` submodule and switch to the tag of the engine version you are working with. For example, `git switch 4.4-stable`.
```

Author: It's a tag so should be "git checkout" and not "switch".

2. If the version is lower than 4.4, you will need to cherry-pick the commit from [here](https://github.com/godotengine/godot/pull/97510 ).  Depending on the version, it might cause a conflict in the file `drivers/vulkan/rendering_context_driver_vulkan.h` which you resolve by taking the changes from the cherry-picked commit for the conflicting section only.

3. Install `scons` if it is not already installed (eg `brew install scons`).

> 4. [Build godot for iOS target](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_ios.html). `scons platform=ios target=editor`

Reviewer: ```suggestion
4. [Build Godot for iOS target](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_ios.html): `scons platform=ios target=editor`
```

> 5. Run plugin creation script. This step need to be repeated for each of the plugins you want to use. Go to the root of this repository then run `./scripts/generate_xcframework.sh <plugin_name> <debug|release|release_debug> 4.0`. The result will be an xcframework in bin directory.

Reviewer: ```suggestion
5. Run the plugin creation script. This step needs to be repeated for each of the plugins you want to use. Go to the root of this repository then run `./scripts/generate_xcframework.sh <plugin_name> <debug|release|release_debug> 4.4`. The result will be an xcframework in bin directory.
```

Author: See scons configuration file, the two options are `3.x` and `4.0`.

### For Godot 4.x:

1. After cloning this repository, go to `godot` submodule and switch to the tag of the engine version you are working with. For example, `git checkout 4.4-stable`.

2. If the version is lower than 4.4, you will need to cherry-pick the commit from [here](https://github.com/godotengine/godot/pull/97510). Depending on the version, it might cause a conflict in the file `drivers/vulkan/rendering_context_driver_vulkan.h` which you resolve by taking the changes from the cherry-picked commit for the conflicting section only.

3. Install `scons` if it is not already installed (eg `brew install scons`).

> 4. [Build Godot for iOS target](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_ios.html): `scons platform=ios target=editor`

Reviewer: Update link to avoid redirect:

```suggestion
4. [Build Godot for iOS target](https://docs.godotengine.org/en/stable/engine_details/development/compiling/compiling_for_ios.html): `scons platform=ios target=editor`
```

### For Godot 4.x:

1. After cloning this repository, go to `godot` submodule and switch to the tag of the engine version you are working with. For example, `git checkout 4.4-stable`.

> 2. If the version is lower than 4.4, you will need to cherry-pick the commit from [here](https://github.com/godotengine/godot/pull/97510 ).  Depending on the version, it might cause a conflict in the file `drivers/vulkan/rendering_context_driver_vulkan.h` which you resolve by taking the changes from the cherry-picked commit for the conflicting section only.

Reviewer: ```suggestion
2. If the version is lower than 4.4, you will need to cherry-pick the commit from [here](https://github.com/godotengine/godot/pull/97510). Depending on the version, it might cause a conflict in the file `drivers/vulkan/rendering_context_driver_vulkan.h` which you resolve by taking the changes from the cherry-picked commit for the conflicting section only.
```

3. Install `scons` if it is not already installed (eg `brew install scons`).

4. [Build Godot for iOS target](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_ios.html): `scons platform=ios target=editor`

5. Run the plugin creation script. This step needs to be repeated for each of the plugins you want to use. Go to the root of this repository then run `./scripts/generate_xcframework.sh <plugin_name> <debug|release|release_debug> 4.0`. The result will be an xcframework in bin directory.

> 

Reviewer: There are some extraneous blank lines here:

```suggestion
4. [Build Godot for iOS target](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_ios.html): `scons platform=ios target=editor`

5. Run the plugin creation script. This step needs to be repeated for each of the plugins you want to use. Go to the root of this repository then run `./scripts/generate_xcframework.sh <plugin_name> <debug|release|release_debug> 4.0`. The result will be an xcframework in bin directory.

```

**Note:** Godot's official `debug` export templates are compiled with the `release_debug` target, *not* the `debug` target.

#### Building a `.a` library

- Run `./scripts/generate_static_library.sh <plugin_name> <debug|release|release_debug> <godot_version>`
  to generate `fat` static library with specific configuration.
- The result `.a` binary will be stored in the `bin/` folder.

#### Building a `.xcframework` library

- Run `./scripts/generate_xcframework.sh <plugin_name> <debug|release|release_debug> <godot_version>`
  to generate `xcframework` with specific configuration.
  `xcframework` allows plugin to support both `arm64` device and `arm64` simulator.
- The result `.xcframework` will be stored in the `bin/` folder as well as intermidiate `.a` binaries.

> ## Adding plugins to Godot project

Reviewer: ```suggestion
## Adding plugins to a Godot project
```

More detailed instructions can be found in the [official docs](https://docs.godotengine.org/en/stable/tutorials/platform/ios/ios_plugin.html#loading-and-using-an-existing-plugin).

1. Build both release and debug xcframeworks of the desired plugin.

> 2. In your godot game project, create `red://ios/plugin` and copy there the `<plugin name>.gdip` (from the plugin source directory in this repository) and both xcframework files for the plugin (from bin directory).

Reviewer: ```suggestion
2. In your Godot project, create `res://ios/plugin` and copy there the `<plugin name>.gdip` (from the plugin source directory in this repository) and both `*.xcframework` files for the plugin (from the `bin/` directory).
```

> 3. Tap Project -> Export and select the “iOS” template. In the “Options” tab, scroll to plugins and enable the plugin.

Reviewer: ```suggestion
3. Go to **Project > Export** and select the iOS export preset (or create one if it doesn't exist already). In the **Options** tab, scroll to **Plugins** and enable the plugin.
```

## C\#

It is possible to verify that the plugin's singleton is loaded using `Engine.HasSingleton("InAppStore")` and get access to it with `Engine.GetSingleton("InAppStore")`.

> The return type is always `GodotObject` and the methods which the plugin expose can be called indirectly with `Variant Call(StringName method, params Variant[] @args)` metod of `GodotObject`.

Reviewer: ```suggestion
The return type is always `GodotObject`. The methods which the plugin expose can be called indirectly with `Variant Call(StringName method, params Variant[] @args)` metod of `GodotObject`.
```
